big_vision/utils.py

```diff
@@ -685,44 +755,84 @@ def recover_tree(keys, values):
   return tree
 
 
+def steps(prefix, config, data_size=None, batch_size=None, default=ValueError):
+  """Gets duration named `prefix` out of `config` and converts it to steps.
+
+  Using this function to access a configuration value that denotes some kind
+  of duration (eg training time, warmup, checkpoint frequency, ...) allows the
+  duration to be specified in terms of steps, epochs, or examples, and converts
+  any of these into steps, such that the training code only deals with steps.
+
+  Args:
+    prefix: The name of the duration to query. The actual config fields can
+      then be one of `prefix_steps`, `prefix_examples`, or `prefix_epochs`.
+    config: The dictionary (config) from which to read the duration.
+    data_size: The total number of training examples in one epoch.
+    batch_size: The number of examples processed per step.
+    default: The default value to return when no duration of the name `prefix`
+      is found in the `config`. Set to `ValueError` (the default) to raise an
+      error instead of returning a default value.
+
+  Returns:
+    The number of steps from the config, or the default value.
+
+  Raises:
+    ValueError if there is no such duration in the config and no default is set.
+  """
+  # Be helpful and make sure only one of _steps, _epochs, _examples is defined.
+  msg = f"Only one of {prefix}_(steps,examples,epochs) should be defined."
+  assert (int(f"{prefix}_steps" in config) +
+          int(f"{prefix}_examples" in config) +
+          int(f"{prefix}_epochs" in config)) <= 1, msg
+
+  # Boy do I anticipate the walrus operator...
+  if f"{prefix}_steps" in config:
+    return config[f"{prefix}_steps"]
+
+  if batch_size and f"{prefix}_examples" in config:
+    return config[f"{prefix}_examples"] // batch_size
+
+  if batch_size and data_size and f"{prefix}_epochs" in config:
+    steps_per_epoch = data_size / batch_size
+    return int(config[f"{prefix}_epochs"] * steps_per_epoch)
+
+  if default is ValueError:
+    raise ValueError(
+        f"Cannot convert {prefix} to steps, due to missing batch_size "
+        f"({batch_size}), data_size ({data_size}), or corresponding entry in "
+        "config:\n" + "\n".join(config.keys()))
+
+  return default
+
+
 def create_learning_rate_schedule(
```

Comment:
fPushing this update means the distillation and the image_text trainers will not work anymore, and even your uvim trainers should not actually be working with this.
See call to `bv_optax.make` in all of them vs in train.py. global_batch_size -> batch_size, and also -steps_per_epoch but +data_size.

Ideally you could change these accordingly, I could do that too, but I'm really short on time until Monday evening.

Reply:
@lucasb-eyer - I noticed the same thing, but then looked in details I realised that the current state is messy and error-prone but not broken at head. I suggest we fix it after this and not try to roll the fix with this one. Alternative lets fix it and do a pull-request before uvim is added.

In detail, note:
 - The old global_batch_size was only need in utils.create_learning_rate_schedule when scale_with_batchsize is set to True. This never happens in any submitted config. This is now replaced by a use of batch_size.. but again those lines don't trigger in any subbmitted config.
 - The old "steps_per_epoch" was used to compute "warmup/cooldown_epochs" which was not used by any config. It is now replaced by a function which depends on the new batch_size and data_size.. but again those are only needed when one has a config with "warmup/cooldown_{epochs/examples}" which is also not used in any config.

Reply:
Yep, your analysis is exactly correct. OK either way for me to fix this after, or before. It's really a simple change, I might quickly make it internally so you can squeeze it in here, but no need to wait for it.